### PR TITLE
LSTM layer for new graph engine. 

### PR DESCRIPTION
**Pull request by Abdurrahheem**
Merge with extra: https://github.com/opencv/opencv_extra/pull/1218

This PR updates/creates LSTM layer compatible with new graph engine. It is based on previous LSTM implementation with some modification on how initializers blobs are processed.

Note: Following tests are currently are disabled 

Two following two tests are disbled since ONNNRuntime does not support `layout=1` attiribute inference. See a detailed issue #26456 on this.
- `LSTM_layout_seq` 
- `LSTM_layout_batch`

Following test fails with the new engine as it is not able to deal with shapes of the form [?, C, H, W]
- `LSTM_Activations`

Works:
- [x] One directional case any batch type 
 - [x] Fix directional case when batch size large than 1
 - [x] Add peepholes attribute

TODO with the next PRs:
 - [ ] Activation support

Note: 
  

> Currently `LSTM_layout_seq`, `LSTM_layout_batch` are disabled as the tests are incorrect. They do not comply with the ONNX standard. Particularly test outputs are of incorrect dimensionality. They produce 3-dimentinal output instead of 4-dimentional. 

------

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/layers/recurrent2_layers.cpp [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
```
Comment (asmorkalov) [inline modules/dnn/src/layers/recurrent2_layers.cpp:41]: Please use modern license header, e.g. https://github.com/opencv/opencv/blob/4.x/modules/dnn/src/backend.cpp

**Comment by asmorkalov** [comment]
`Test_ONNX_layers.LSTM_cell_with_peepholes/0, where GetParam() = OCV/CPU` Crashes with sigsegv on my Ubuntu 24.04:
```
[ RUN      ] Test_ONNX_layers.LSTM_cell_with_peepholes/0, where GetParam() = OCV/CPU
basename: lstm_cell_with_peepholes

==>parseLSTM
initial proto node output size: 2
output [0] is empty: 1
output [1] is empty: 0
needs_yc: 0
needs_yh: 1
needs_y: 0
output [0] name: 
output [1] name: Y_h
before addLayer num proto outputs: 2

==>addLayer
node outputs size: 2
==>addLayer done

==>parseLSTM done


==>LSTM getMemoryShapes
input[0] shape: [1 x 2 x 4]
input[1] shape: [1 x 12 x 4]
input[2] shape: [1 x 12 x 3]
input[3] shape: [1 x 24]
input[4] shape: [2]
input[5] shape: [1 x 2 x 3]
input[6] shape: [1 x 2 x 3]
input[7] shape: [1 x 9]
ninputs: 8
noutputs: 2
_hidSize: 3
_inpSize: 4
_batchSize: 2
outResShape: [1 x 1 x 2 x 3]
produceCellOutput: 0
produceOutputYh: 1
output[0] shape: [1 x 1 x 2 x 3]
output[1] shape: [1 x 2 x 3]
internal[0] shape: [2 x 3]
internal[1] shape: [2 x 3]
internal[2] shape: [2 x 1]
internal[3] shape: [2 x 12]
==>LSTM getMemoryShapes done

==>noutputs: 2
==>outShapes size: 2
==>tempShapes size: 4

==>LSTM getTypes
==>LSTM getTypes done

[ WARN:0@0.048] global net_impl_backend.cpp:229 setPreferableTarget Targtes are not supported by the new graph egine for now
==>noutputs: 2
==>outShapes size: 0

==>LSTM getMemoryShapes
input[0] shape: [1 x 2 x 4]
input[1] shape: [1 x 12 x 4]
input[2] shape: [1 x 12 x 3]
input[3] shape: [1 x 24]
input[4] shape: [2]
input[5] shape: [1 x 2 x 3]
input[6] shape: [1 x 2 x 3]
input[7] shape: [1 x 9]
ninputs: 8
noutputs: 2
_hidSize: 3
_inpSize: 4
_batchSize: 2
outResShape: [1 x 1 x 2 x 3]
produceCellOutput: 0
produceOutputYh: 1
output[0] shape: [1 x 1 x 2 x 3]
output[1] shape: [1 x 2 x 3]
internal[0] shape: [2 x 3]
internal[1] shape: [2 x 3]
internal[2] shape: [2 x 1]
internal[3] shape: [2 x 12]
==>LSTM getMemoryShapes done

==>noutputs: 2
==>tempShapes size: 4

==>LSTM getTypes
==>LSTM getTypes done

==>outShapes size: 2
==>outTypes size: 2

==>LSTM forward
input shape: 1 x 2 x 4
input sum: 36
input shape: 1 x 12 x 4
input sum: 4.8
input shape: 1 x 12 x 3
input sum: 3.6
input shape: 1 x 24
input sum: 0
input shape: 2
input sum: 2
input shape: 1 x 2 x 3
input sum: 0
input shape: 1 x 2 x 3
input sum: 0
input shape: 1 x 9
input sum: 0.9
internal shape: 2 x 3
internal shape: 2 x 3
internal shape: 2 x 1
internal shape: 2 x 12
output shape: 1 x 1 x 2 x 3
output shape: 1 x 2 x 3
transforming blobs
transformBlobs
blobs[0] shape: 5
Wx shape: 1 x 12 x 4
Wh shape: 1 x 12 x 3
Wx shape: 12 x 4
Wh shape: 12 x 3
b shape: 1 x 12
blobs[0] shape: 12 x 4
blobs[1] shape: 12 x 3
blobs[2] shape: 1 x 12
*********************
Wx sum: 4.8
Wh sum: 3.6
bias sum: 0
h_0 sum: 0
c_0 sum: 0
Wx shape: 12 x 4
Wh shape: 12 x 3
bias shape: 1 x 12
h_0 shape: 2 x 3
c_0 shape: 2 x 3
Wx sum: 4.8
Wh sum: 3.6
bias sum: 0
h_0 sum: 0
c_0 sum: 0
Wx shape: 12 x 4
Wh shape: 12 x 3
bias shape: 1 x 12
h_0 shape: 2 x 3
c_0 shape: 2 x 3
input[0] shape: 1 x 2 x 4
seqLenth: 1
batchSize: 2
xTs shape: 2 x 4
hOutTs shape: 2 x 3
hInternal shape: 2 x 3
cInternal shape: 2 x 3
dummyOnes shape: 2 x 1
gates shape: 2 x 12
xTs sum: [36, 0, 0, 0]
hOutTs sum: [0, 0, 0, 0]
hInternal sum: [0, 0, 0, 0]
cInternal sum: [0, 0, 0, 0]
dummyOnes sum: [2, 0, 0, 0]
gates sum: [0, 0, 0, 0]
cOut shape: 
cOut shape: 
cOutTs shape: 
batchSize: 2
seqLenth: 1
========================
 ------------------ 
ts: 0
xCurr shape: 2 x 4
xCurr sum: [36, 0, 0, 0]
Wx shape: 12 x 4
Wx sum: 4.8
gates shape: 2 x 12
gates sum: 0
xCurr: [1, 2, 3, 4;
 5, 6, 7, 8]
Wx: [0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1;
 0.1, 0.1, 0.1, 0.1]
gates sum: 43.2
dummyOnes shape: 2 x 1
bias shape: 1 x 12
gates sum: [43.2, 0, 0, 0]
hInternal shape: 2 x 3
Wh shape: 12 x 3
gates shape: 2 x 12
gates sum: [43.2, 0, 0, 0]
gateI shape: 2 x 3
gateF shape: 2 x 3
gateO shape: 2 x 3
gateG shape: 2 x 3
curRowRange: [0 : 2)
produceCellOutput: 0
cInternal shape: 2 x 3
cOutTs shape: 
cInternal sum: [4.43225, 0, 0, 0]
cOutTs sum: [0, 0, 0, 0]
hInternal shape: 2 x 3
hOutTs shape: 2 x 3
hInternal sum: [3.1368, 0, 0, 0]
hOutTs sum: [3.1368, 0, 0, 0]
cOutTs type: 0
cInternal type: 5
cOutTs sum: [0, 0, 0, 0]
 ===> 
produceOutputYh needs to be produced
Segmentation fault (core dumped)
```

**Comment by Abdurrahheem** [comment]
> `Test_ONNX_layers.LSTM_cell_with_peepholes/0, where GetParam() = OCV/CPU` Crashes with sigsegv on my Ubuntu 24.04:
> 
> ```
> [ RUN      ] Test_ONNX_layers.LSTM_cell_with_peepholes/0, where GetParam() = OCV/CPU
> basename: lstm_cell_with_peepholes
> 
> ==>parseLSTM
> initial proto node output size: 2
> output [0] is empty: 1
> output [1] is empty: 0
> needs_yc: 0
> needs_yh: 1
> needs_y: 0
> output [0] name: 
> output [1] name: Y_h
> before addLayer num proto outputs: 2
> 
> ==>addLayer
> node outputs size: 2
> ==>addLayer done
> 
> ==>parseLSTM done
> 
> 
> ==>LSTM getMemoryShapes
> input[0] shape: [1 x 2 x 4]
> input[1] shape: [1 x 12 x 4]
> input[2] shape: [1 x 12 x 3]
> input[3] shape: [1 x 24]
> input[4] shape: [2]
> input[5] shape: [1 x 2 x 3]
> input[6] shape: [1 x 2 x 3]
> input[7] shape: [1 x 9]
> ninputs: 8
> noutputs: 2
> _hidSize: 3
> _inpSize: 4
> _batchSize: 2
> outResShape: [1 x 1 x 2 x 3]
> produceCellOutput: 0
> produceOutputYh: 1
> output[0] shape: [1 x 1 x 2 x 3]
> output[1] shape: [1 x 2 x 3]
> internal[0] shape: [2 x 3]
> internal[1] shape: [2 x 3]
> internal[2] shape: [2 x 1]
> internal[3] shape: [2 x 12]
> ==>LSTM getMemoryShapes done
> 
> ==>noutputs: 2
> ==>outShapes size: 2
> ==>tempShapes size: 4
> 
> ==>LSTM getTypes
> ==>LSTM getTypes done
> 
> [ WARN:0@0.048] global net_impl_backend.cpp:229 setPreferableTarget Targtes are not supported by the new graph egine for now
> ==>noutputs: 2
> ==>outShapes size: 0
> 
> ==>LSTM getMemoryShapes
> input[0] shape: [1 x 2 x 4]
> input[1] shape: [1 x 12 x 4]
> input[2] shape: [1 x 12 x 3]
> input[3] shape: [1 x 24]
> input[4] shape: [2]
> input[5] shape: [1 x 2 x 3]
> input[6] shape: [1 x 2 x 3]
> input[7] shape: [1 x 9]
> ninputs: 8
> noutputs: 2
> _hidSize: 3
> _inpSize: 4
> _batchSize: 2
> outResShape: [1 x 1 x 2 x 3]
> produceCellOutput: 0
> produceOutputYh: 1
> output[0] shape: [1 x 1 x 2 x 3]
> output[1] shape: [1 x 2 x 3]
> internal[0] shape: [2 x 3]
> internal[1] shape: [2 x 3]
> internal[2] shape: [2 x 1]
> internal[3] shape: [2 x 12]
> ==>LSTM getMemoryShapes done
> 
> ==>noutputs: 2
> ==>tempShapes size: 4
> 
> ==>LSTM getTypes
> ==>LSTM getTypes done
> 
> ==>outShapes size: 2
> ==>outTypes size: 2
> 
> ==>LSTM forward
> input shape: 1 x 2 x 4
> input sum: 36
> input shape: 1 x 12 x 4
> input sum: 4.8
> input shape: 1 x 12 x 3
> input sum: 3.6
> input shape: 1 x 24
> input sum: 0
> input shape: 2
> input sum: 2
> input shape: 1 x 2 x 3
> input sum: 0
> input shape: 1 x 2 x 3
> input sum: 0
> input shape: 1 x 9
> input sum: 0.9
> internal shape: 2 x 3
> internal shape: 2 x 3
> internal shape: 2 x 1
> internal shape: 2 x 12
> output shape: 1 x 1 x 2 x 3
> output shape: 1 x 2 x 3
> transforming blobs
> transformBlobs
> blobs[0] shape: 5
> Wx shape: 1 x 12 x 4
> Wh shape: 1 x 12 x 3
> Wx shape: 12 x 4
> Wh shape: 12 x 3
> b shape: 1 x 12
> blobs[0] shape: 12 x 4
> blobs[1] shape: 12 x 3
> blobs[2] shape: 1 x 12
> *********************
> Wx sum: 4.8
> Wh sum: 3.6
> bias sum: 0
> h_0 sum: 0
> c_0 sum: 0
> Wx shape: 12 x 4
> Wh shape: 12 x 3
> bias shape: 1 x 12
> h_0 shape: 2 x 3
> c_0 shape: 2 x 3
> Wx sum: 4.8
> Wh sum: 3.6
> bias sum: 0
> h_0 sum: 0
> c_0 sum: 0
> Wx shape: 12 x 4
> Wh shape: 12 x 3
> bias shape: 1 x 12
> h_0 shape: 2 x 3
> c_0 shape: 2 x 3
> input[0] shape: 1 x 2 x 4
> seqLenth: 1
> batchSize: 2
> xTs shape: 2 x 4
> hOutTs shape: 2 x 3
> hInternal shape: 2 x 3
> cInternal shape: 2 x 3
> dummyOnes shape: 2 x 1
> gates shape: 2 x 12
> xTs sum: [36, 0, 0, 0]
> hOutTs sum: [0, 0, 0, 0]
> hInternal sum: [0, 0, 0, 0]
> cInternal sum: [0, 0, 0, 0]
> dummyOnes sum: [2, 0, 0, 0]
> gates sum: [0, 0, 0, 0]
> cOut shape: 
> cOut shape: 
> cOutTs shape: 
> batchSize: 2
> seqLenth: 1
> ========================
>  ------------------ 
> ts: 0
> xCurr shape: 2 x 4
> xCurr sum: [36, 0, 0, 0]
> Wx shape: 12 x 4
> Wx sum: 4.8
> gates shape: 2 x 12
> gates sum: 0
> xCurr: [1, 2, 3, 4;
>  5, 6, 7, 8]
> Wx: [0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1;
>  0.1, 0.1, 0.1, 0.1]
> gates sum: 43.2
> dummyOnes shape: 2 x 1
> bias shape: 1 x 12
> gates sum: [43.2, 0, 0, 0]
> hInternal shape: 2 x 3
> Wh shape: 12 x 3
> gates shape: 2 x 12
> gates sum: [43.2, 0, 0, 0]
> gateI shape: 2 x 3
> gateF shape: 2 x 3
> gateO shape: 2 x 3
> gateG shape: 2 x 3
> curRowRange: [0 : 2)
> produceCellOutput: 0
> cInternal shape: 2 x 3
> cOutTs shape: 
> cInternal sum: [4.43225, 0, 0, 0]
> cOutTs sum: [0, 0, 0, 0]
> hInternal shape: 2 x 3
> hOutTs shape: 2 x 3
> hInternal sum: [3.1368, 0, 0, 0]
> hOutTs sum: [3.1368, 0, 0, 0]
> cOutTs type: 0
> cInternal type: 5
> cOutTs sum: [0, 0, 0, 0]
>  ===> 
> produceOutputYh needs to be produced
> Segmentation fault (core dumped)
> ```

Peepholes not yet. Please not run yet

**Comment by asmorkalov** [comment]
@there is some strange bug or shared state in tests. E.g. if I run all cases for LSTM like `./bin/opencv_test_dnn --gtest_filter=*LSTM*`
I see 
```
[  FAILED  ] 20 tests, listed below:
[  FAILED  ] Test_ONNX_layers.LSTM_hidden/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_layers.LSTM_hidden/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_hidden/2, where GetParam() = OCV/CPU
[  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/2, where GetParam() = OCV/CPU
[  FAILED  ] Test_ONNX_layers.LSTM_cell_bidirectional/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_layers.LSTM_cell_bidirectional/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_cell_bidirectional/2, where GetParam() = OCV/CPU
[  FAILED  ] Test_ONNX_layers.LSTM_cell_with_peepholes/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_layers.LSTM_cell_with_peepholes/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_cell_with_peepholes/2, where GetParam() = OCV/CPU
[  FAILED  ] Test_ONNX_layers.LSTM_cell_batchsize_1_seqlen_50/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_cell_batchsize_5_seqlen_5/2, where GetParam() = OCV/CPU
[  FAILED  ] Test_ONNX_layers.LSTM_layout_seq/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_layers.LSTM_layout_seq/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_layout_seq/2, where GetParam() = OCV/CPU
[  FAILED  ] Test_ONNX_layers.LSTM_layout_batch/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_layers.LSTM_layout_batch/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_layout_batch/2, where GetParam() = OCV/CPU
```
But if I run cases for hidden state, e.g. `./bin/opencv_test_dnn --gtest_filter=Test_ONNX_layers.LSTM_hidden*`, I see
```
[==========] 6 tests from 1 test case ran. (20 ms total)
[  PASSED  ] 3 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/2, where GetParam() = OCV/CPU
```

It means, that `Test_ONNX_layers.LSTM_hidden` passes.

**Comment by asmorkalov** [comment]
"Test_ONNX_layers.LSTM_hidden/0, where GetParam() = OCV/CPU" always fails in debug build.

**Comment by Abdurrahheem** [comment]
> @there is some strange bug or shared state in tests. E.g. if I run all cases for LSTM like `./bin/opencv_test_dnn --gtest_filter=*LSTM*` I see
> 
> ```
> [  FAILED  ] 20 tests, listed below:
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden/0, where GetParam() = OCV/OCL
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden/2, where GetParam() = OCV/CPU
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/0, where GetParam() = OCV/OCL
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/2, where GetParam() = OCV/CPU
> [  FAILED  ] Test_ONNX_layers.LSTM_cell_bidirectional/0, where GetParam() = OCV/OCL
> [  FAILED  ] Test_ONNX_layers.LSTM_cell_bidirectional/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.LSTM_cell_bidirectional/2, where GetParam() = OCV/CPU
> [  FAILED  ] Test_ONNX_layers.LSTM_cell_with_peepholes/0, where GetParam() = OCV/OCL
> [  FAILED  ] Test_ONNX_layers.LSTM_cell_with_peepholes/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.LSTM_cell_with_peepholes/2, where GetParam() = OCV/CPU
> [  FAILED  ] Test_ONNX_layers.LSTM_cell_batchsize_1_seqlen_50/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.LSTM_cell_batchsize_5_seqlen_5/2, where GetParam() = OCV/CPU
> [  FAILED  ] Test_ONNX_layers.LSTM_layout_seq/0, where GetParam() = OCV/OCL
> [  FAILED  ] Test_ONNX_layers.LSTM_layout_seq/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.LSTM_layout_seq/2, where GetParam() = OCV/CPU
> [  FAILED  ] Test_ONNX_layers.LSTM_layout_batch/0, where GetParam() = OCV/OCL
> [  FAILED  ] Test_ONNX_layers.LSTM_layout_batch/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.LSTM_layout_batch/2, where GetParam() = OCV/CPU
> ```
> 
> But if I run cases for hidden state, e.g. `./bin/opencv_test_dnn --gtest_filter=Test_ONNX_layers.LSTM_hidden*`, I see
> 
> ```
> [==========] 6 tests from 1 test case ran. (20 ms total)
> [  PASSED  ] 3 tests.
> [  FAILED  ] 3 tests, listed below:
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/0, where GetParam() = OCV/OCL
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.LSTM_hidden_bidirectional/2, where GetParam() = OCV/CPU
> ```
> 
> It means, that `Test_ONNX_layers.LSTM_hidden` passes.

this behavior is only reproducable on Ubuntu. Not sure why

**Comment by asmorkalov** [comment]
@vpisarev @dkurt Could you take a look?

File: modules/dnn/src/layers/recurrent2_layers.cpp [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+        void transformBlobs(std::vector<Mat>& blobs)
+        {
+            Mat &Wx = blobs[0];
+            Mat &Wh = blobs[1];
+            Mat &b = blobs[2];
+
+            // std::vector<Mat> cudaWorkaround;
+            // cudaWorkaround.push_back(Wx.clone());
+            // cudaWorkaround.push_back(Wh.clone());
+            // cudaWorkaround.push_back(b.clone());
```
Comment (asmorkalov) [inline modules/dnn/src/layers/recurrent2_layers.cpp:527]: Please remove if not relevant.

File: modules/dnn/test/test_onnx_importer.cpp [hunk 3]
```diff
@@ -1328,7 +1328,7 @@ TEST_P(Test_ONNX_layers, Split_EltwiseMax)
     testONNXModels("split_max");
 }
 
-TEST_P(Test_ONNX_layers, LSTM_Activations)
```
Comment (asmorkalov) [inline modules/dnn/test/test_onnx_importer.cpp:1331]: Please add note why the test is disabled. Here and bellow.

File: modules/dnn/src/layers/recurrent2_layers.cpp [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+        }
+
+        void getTypes(const std::vector<MatType>& inputs,
+                      const int requiredOutputs,
+                      const int requiredInternals,
+                      std::vector<MatType>& outputs,
+                      std::vector<MatType>& internals) const CV_OVERRIDE
+        {
+            outputs.assign(requiredOutputs, inputs[0]);
+            internals.assign(4, inputs[0]);
```
Comment (asmorkalov) [inline modules/dnn/src/layers/recurrent2_layers.cpp:235]: According to implementation, only fp32 and m.b. fp64 are supported. Need ti add check/assert here.

File: modules/dnn/src/layers/recurrent2_layers.cpp [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+            inputs_arr.getMatVector(input);
+            outputs_arr.getMatVector(output);
+            internals_arr.getMatVector(internals);
+
+            // set outputs to 0
+            for (auto& out : output)
+                out.setTo(0);
+
+            // transform weights matrices similar to old LSTM parser
+            std::vector<Mat> blobs_ = {input[1], input[2], input[3], input[5], input[6]};
```
Comment (asmorkalov) [inline modules/dnn/src/layers/recurrent2_layers.cpp:254]: Please add check that the layer got enough inputs.

File: modules/dnn/src/layers/recurrent2_layers.cpp [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+            if (usePeephole){
+                blobs_.push_back(input[7]);
```
Comment (asmorkalov) [inline modules/dnn/src/layers/recurrent2_layers.cpp:256]: The same.

File: modules/dnn/src/layers/recurrent2_layers.cpp [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+
+                int shp[] = {1, part1.size[2] * part1.size[3]};
+                part1 = part1.reshape(1, sizeof(shp)/sizeof(shp[0]), shp);
+                part2 = part2.reshape(1, sizeof(shp)/sizeof(shp[0]), shp);
+
+                vconcat(part1, part2, dst);
+
+                int finalShape[] = {2, batchSize, numHidden};
+                dst = dst.reshape(1, sizeof(finalShape)/sizeof(finalShape[0]), finalShape);
+
```
Comment (asmorkalov) [inline modules/dnn/src/layers/recurrent2_layers.cpp:463]: redundant space line.

**Comment by asmorkalov** [comment]
Works well on my side.

File: modules/dnn/src/init.cpp [hunk 8]
```diff
@@ -209,6 +209,7 @@ void initializeLayerFactory()
     CV_DNN_REGISTER_LAYER_CLASS(FlowWarp,       FlowWarpLayer);
 
     CV_DNN_REGISTER_LAYER_CLASS(LSTM,           LSTMLayer);
+    CV_DNN_REGISTER_LAYER_CLASS(LSTM2,          LSTM2Layer);
```
Comment (dkurt) [inline modules/dnn/src/init.cpp:212]: That is the first layer named in such a way. I think that layer name should remain the same while `Impl` class should be different for different versions of the engine

File: modules/dnn/test/test_onnx_importer.cpp (hunk 3 above)
Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:1331]: As for me, layer name should be `LSTM` but switch implementation class depends on engine settings.

File: modules/dnn/src/init.cpp (hunk 8 above)
Comment (Abdurrahheem) [reply modules/dnn/src/init.cpp:212]: Not first. here are more examples [[1](https://github.com/opencv/opencv/blob/979428d5908b5396237b18cbd50d0b18265e2463/modules/dnn/src/init.cpp#L92)], [[2](https://github.com/opencv/opencv/blob/979428d5908b5396237b18cbd50d0b18265e2463/modules/dnn/src/init.cpp#L88)], [[3](https://github.com/opencv/opencv/blob/979428d5908b5396237b18cbd50d0b18265e2463/modules/dnn/src/init.cpp#L107)]

Comment (dkurt) [reply modules/dnn/src/init.cpp:212]: Oh, I see.

**Review by asmorkalov — APPROVED** [review]

File: modules/dnn/test/test_onnx_importer.cpp (hunk 3 above)
Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:1331]: Is that comment still actual? To understand why working tests has been disabled and no new alternatives added.

**Comment by dkurt** [comment]
@asmorkalov, wanted to clarify about the disabled tests

**Comment by asmorkalov** [comment]
```
[ RUN      ] Test_Model.TextRecognition/2, where GetParam() = OCV/CPU
[ WARN:0@0.140] global net_impl2.cpp:951 tryInferGraphShapes DNN/InferShape: Layer 'Reshape_42' (Reshape2) output shapes cannot be inferenced without running forward()
[ WARN:0@0.140] global net_impl_backend.cpp:229 setPreferableTarget Targtes are not supported by the new graph egine for now
/mnt/Projects/Projects/opencv/modules/dnn/test/test_model.cpp:154: Failure
Expected equality of these values:
  results[0]
    Which is: "ewelomae"
  seq
    Which is: "welcome"
ROI[0]: /mnt/Projects/Projects/opencv_extra/testdata/dnn/text_rec_test.png
[  FAILED  ] Test_Model.TextRecognition/2, where GetParam() = OCV/CPU (410 ms)
```

**Comment by asmorkalov** [comment]
Please tune shape check:
```
[ RUN      ] Test_ONNX_layers.LSTM_layout_batch/0, where GetParam() = OCV/CPU
[ WARN:0@53.002] global net_impl_backend.cpp:229 setPreferableTarget Targets are not supported by the new graph engine for now
/home/ci/opencv/modules/dnn/test/test_onnx_importer.cpp:146: Failure
Expected equality of these values:
  ref.shape()
    Which is: 52-byte object <03-00 00-00 00-00 00-00 00-00 00-00 01-00 00-00 05-00 00-00 07-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00>
  out.shape()
    Which is: 52-byte object <04-00 00-00 00-00 00-00 00-00 00-00 05-00 00-00 02-00 00-00 01-00 00-00 07-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00>
shape1: [1 x 5 x 7]
shape2: [5 x 2 x 1 x 7]
unknown file: Failure
C++ exception with description "OpenCV(5.0.0-pre) /home/ci/opencv/modules/ts/src/ts_func.cpp:1587: error: (-215:Assertion failed) shape1 == shape2 && "shapes of compared arrays must be the same" in function 'norm'
" thrown in the test body.
[  FAILED  ] Test_ONNX_layers.LSTM_layout_batch/0, where GetParam() = OCV/CPU (1 ms)
```

**Comment by asmorkalov** [comment]
@dkurt @vpisarev Could you take a look on the PR again. I want to merge it today.

File: modules/dnn/test/test_layers.cpp [hunk 9]
```diff
@@ -2740,4 +2740,54 @@ INSTANTIATE_TEST_CASE_P(TestLayerFusion, ConvolutionActivationEltwiseFusion, Com
                         TestLayerFusion::dnnBackendsAndTargetsForFusionTests()
 ));
 
+TEST(Layer_LSTM, repeatedInference)
+{
+    std::string onnx_file_path = findDataFile("dnn/onnx/models/onnxscript_lstm.onnx", false);
```
Comment (opencv-alalek) [inline modules/dnn/test/test_layers.cpp:2745]: Merged against RED CI:

```
[----------] 1 test from Layer_LSTM
[ RUN      ] Layer_LSTM.repeatedInference
TEST ERROR: Don't use 'optional' findData() for dnn/onnx/models/onnxscript_lstm.onnx
unknown file: Failure
C++ exception with description "OpenCV(5.0.0-pre) /build/precommit_linux64/5.x/opencv/modules/ts/src/ts.cpp:969: error: (-215:Assertion failed) required || result_.empty() in function 'findData'
" thrown in the test body.
[  FAILED  ] Layer_LSTM.repeatedInference (0 ms)
```

**Pull request merged**